### DISCUSSION TITLE
API - Help with an error

### DISCUSSION THREAD

**Discussion by bigtimemimi** [cloudbet/docs#39] [Q&A]
Hi and thank you for your time

I'm by no means a software-savvy person, I'm only a novice looking to have fun with some betting-related projects. Therefore, I might be lacking some pivotal knowledge that I'm not even aware of. 

Long story short, I've created a Python bot that grabs soccer odds from Cloudbet, compares them within the given time frame, and attempts to place a bet if a given condition is met. 

1) I've created a loop but after some unspecified time (sometimes 15 minutes, sometimes 3 hours), the software encounters an error:

```
raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

```
2) I've used the same approach for creating a bot that scans 1X2, O/U, and AH markets. Is it okay to run the aforementioned 3 scripts parallelly?

3) For reasons unknown to me, database odds values get skewed in ~10% scans, so bot attempts to place a bet on a match and number that is higher than current odds. I don't want to spam API with requests, and since I'm not sure how to fix this, would adding `time.sleep(1)` between attempts be of help, or sending 15-20 bet attempts isn't a problem?

4) I don't want to send too many requests - technically, is this 1 API request? `response = session.get(url, params={"sport": "soccer", "date": date, "limit": "155"})`  Or is that considered 155 requests?

**Comment by kgravenreuth** [comment]
Hi @bigtimemimi 

While it is tough for us to comment without seeing more details about the code, here are some pointers:

- For live events, markets might be removed (i.e. `markets: {}`) from time to time since live events are volatile and markets could be suspended due to various reasons. 
- We would recommend you to ingest the Feeds API for latest odds, while you should be careful not to spam our Trading (betting) API with too many invalid odds requests. Also, please make sure to use the response of the Trading API endpoint to get more details:

Refer to our [Trading API Documentation for bet status](https://www.cloudbet.com/api/?urls.primaryName=Trading#/Trading/PlaceBet)

```
STAKE_ABOVE_MAX - stake requested was above the current maximum stake on a selection. Please reference stake from response payload about the corrected value for retry
STAKE_BELOW_MIN - stake requested was below the current minimum stake on a selection. Please reference stake from response payload about the corrected value for retry
LIABILITY_LIMIT_EXCEEDED - your current liability limit on this event was exceeded. Please reference stake from response payload about the corrected value for retry
```

We have a rate limit of [1 request per second on the Trading API](https://www.cloudbet.com/api/?urls.primaryName=Trading)

Regards,
Klaus Gravenreuth
Product & Engineering @ Cloudbet
